### PR TITLE
CP-49129: Drop global lock around sexpr parsing

### PR DESCRIPTION
**Pull request by GabrielBuica**
Add @contificate unit test that demonstrates the limitations of  `ocamlyacc`'s concurrency.

Change from `ocamlyacc`  to `menhir`

Drop the global lock around `sexpr` parsing to improve performance.

### CODE REVIEW

**Comment by psafont** [comment]
@GabrielBuica you need to sign off the first commit as well

**Review by psafont — APPROVED** [review]

**Comment by edwintorok** [comment]
There is also https://github.com/xapi-project/xen-api/compare/master...edwintorok:xen-api:private/edvint/sexpr-safe, which was waiting for these tests.

There are a few more ocamlyacc uses, and we should switch those to menhir as well, can you cherry-pick https://github.com/xapi-project/xen-api/commit/041f898056f9b58d76f8432374b32dc17f4b1919

**Comment by edwintorok** [comment]
Maybe also modify `quality-gate.sh` to check with `grep` that there are no uses of `ocamlyacc` anymore to prevent it accidentally being added back in the future (it is part of the standard OCaml distribution, so we cannot remove the actual executable)

**Comment by GabrielBuica** [comment]
I'll move it to draft until I address @edwintorok's comments.

File: ocaml/database/dune
Comment (psafont) [inline ocaml/database/dune:3]: are these wrapped around mutexes as well?

Comment (edwintorok) [reply ocaml/database/dune:3]: Not directly, so either we had a race condition here on master, or we had another mutex that happened to protect this too.

We should probably backport this PR to Yangtze as well in that case, even though the Sexpression parser was safe due to the mutex, this other usage in the database (part of db_actions.ml) was not, and probably all our calls to `get_all_records_where` are not thread-safe/subject to race condition and random failures...

The xenopsd cli one doesn't matter, because that is a single-threaded CLI.

File: quality-gate.sh
Comment (edwintorok) [inline quality-gate.sh:89]: minor typo: usages.

**Comment by edwintorok** [comment]
A xenops test is failing here, is this the same one that you've seen @psafont ?

**Comment by edwintorok** [comment]
It'd be useful to keep/add Colin Barr's (@contificate ?) signoff on https://github.com/xapi-project/xen-api/pull/5682/commits/46a07088da962c754ffcbe56ab54fe54a68bf22b, although the commit message mentions his name, we usually add signoffs if multiple people worked on the same commit (it is fine to keep your signoff there too if you modified it, in this case you moved it into xapi). Did the original commit have no signoff?

**Review by edwintorok — APPROVED** [review]
You're right that git retained authorship information (there is a commit author and committer, and they can be different).

The 'Signed-off-by' is mainly used by the Linux kernel and similar projects, and they recommend adding Signed-off-by as a patch follows its route to mainline, so e.g. people handling the patch can add their own sign-offs: https://www.kernel.org/doc/html/latest/process/submitting-patches.html?highlight=signed%20off

**Review by psafont — APPROVED** [review]

**Pull request merged**
